packages/theme/src/cli/utilities/asset-ignore.ts

```diff
@@ -13,10 +13,23 @@ export function applyIgnoreFilters<T extends {key: string}>(
   const ignoreOptions = options.ignore ?? []
   const onlyOptions = options.only ?? []
 
-  return files
-    .filter(filterBy(shopifyIgnore, '.shopifyignore'))
+  const negatedPatterns = shopifyIgnore.filter((pattern) => pattern.startsWith('!')).map((pattern) => pattern.slice(1))
+  const normalPatterns = shopifyIgnore.filter((pattern) => !pattern.startsWith('!'))
+
+  let filteredFiles = files.filter(filterBy(normalPatterns, '.shopifyignore'))
+
+  if (negatedPatterns.length > 0) {
+    filteredFiles = filteredFiles.concat(files.filter(filterBy(negatedPatterns, '.shopifyignore', true)))
+  }
+
+  filteredFiles = filteredFiles
     .filter(filterBy(ignoreOptions, '--ignore'))
     .filter(filterBy(onlyOptions, '--only', true))
+
+  const uniqueFiles = Array.from(new Set(filteredFiles.map((file) => file.key))).map((key) =>
```

Comment:
I'm on the fence with this. This was an additional guard in the event any of the patterns or regexes somehow added a file twice. We comb through the file list and remove any non-unique values. 

In my tests and 🎩 I couldn't create a scenario where this happened but with how tricky regex matching can be, it could be a possibility.

I'm open to feedback on whether or not to keep this

Reply:
Now that we're doing more operations, wdyt about grouping the logic by source?
- Shopifyignore
- ignore
- only

Reply:
Will implement that now.